src/Compilers/Core/Portable/Serialization/StreamObjectWriter.cs

```diff
@@ -26,14 +25,17 @@ namespace Roslyn.Utilities
     internal sealed partial class StreamObjectWriter : ObjectWriter, IDisposable
     {
         private readonly BinaryWriter _writer;
-        private readonly ObjectBinder _binder;
+        private readonly ObjectBinder _binderOpt;
         private readonly bool _recursive;
         private readonly CancellationToken _cancellationToken;
 
         /// <summary>
-        /// Map of serialized object's reference ids.
+        /// Map of serialized object's reference ids.  The object-reference-map uses refernece equality
+        /// for performance.  While the string-reference-map uses value-equality for greater cache hits 
+        /// and reuse.
         /// </summary>
-        private readonly ReferenceMap _referenceMap;
+        private readonly WriterReferenceMap _objectReferenceMap;
+        private readonly WriterReferenceMap _stringReferenceMap;
```

Comment:
we have specialized reference maps for strings vs objects.  That's because we want to use reference-equality for objects (for perf), but would like value-equality for strings (for compaction).